src/xwayland/xwm/mod.rs

```diff
@@ -0,0 +1,1167 @@
+            _NET_WM_WINDOW_TYPE,
+            _NET_WM_WINDOW_TYPE_DROPDOWN_MENU,
+            _NET_WM_WINDOW_TYPE_DIALOG,
+            _NET_WM_WINDOW_TYPE_MENU,
+            _NET_WM_WINDOW_TYPE_NOTIFICATION,
+            _NET_WM_WINDOW_TYPE_NORMAL,
+            _NET_WM_WINDOW_TYPE_POPUP_MENU,
+            _NET_WM_WINDOW_TYPE_SPLASH,
+            _NET_WM_WINDOW_TYPE_TOOLBAR,
+            _NET_WM_WINDOW_TYPE_TOOLTIP,
+            _NET_WM_WINDOW_TYPE_UTILITY,
+            _NET_WM_STATE_MODAL,
+            _MOTIF_WM_HINTS,
+
+            // server -> client
+            WM_S0,
+            WM_STATE,
+            _NET_WM_CM_S0,
+            _NET_SUPPORTED,
+            _NET_ACTIVE_WINDOW,
+            _NET_CLIENT_LIST,
+            _NET_CLIENT_LIST_STACKING,
+            _NET_WM_PING,
+            _NET_WM_STATE,
+            _NET_WM_STATE_MAXIMIZED_VERT,
+            _NET_WM_STATE_MAXIMIZED_HORZ,
+            _NET_WM_STATE_HIDDEN,
+            _NET_WM_STATE_FULLSCREEN,
+            _NET_WM_STATE_FOCUSED,
+            _NET_SUPPORTING_WM_CHECK,
+        }
+    }
+}
+pub use self::atoms::Atoms;
+
+crate::utils::ids::id_gen!(next_xwm_id, XWM_ID, XWM_IDS);
+
+/// Id of an X11 WM
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct XwmId(usize);
+
+/// Handler trait for X11WM interactions
+pub trait XwmHandler {
+    /// [`X11WM`] getter for a given ID.
+    fn xwm_state(&mut self, xwm: XwmId) -> &mut X11WM;
+
+    /// A new X11 window was created.
+    ///
+    /// New windows are not mapped yet, but various information is already accessible.
+    /// In general new windows will either stay in this state, if they serve secondary purposes
+    /// or request to be mapped shortly afterwards.
+    fn new_window(&mut self, xwm: XwmId, window: X11Surface);
+    /// A new X11 window with the override redirect flag.
+    ///
+    /// New override_redirect windows are not mapped yet, but can become any time.
+    /// Window manager are not supposed to manage these windows and thus cannot intercept
+    /// most operations (including mapping).
+    ///
+    /// It is best to replicate their state in smithay as faithfully as possible (e.g. positioning)
+    /// and don't touch their state in any way.
+    fn new_override_redirect_window(&mut self, xwm: XwmId, window: X11Surface);
+    /// Window requests to be mapped.
+    ///
+    /// To grant the wish you have to call `X11Surface::set_mapped(true)` for the window to become visible.
+    fn map_window_request(&mut self, xwm: XwmId, window: X11Surface);
+    /// Override redirect window was mapped.
+    ///
+    /// This is a notification. The XWM cannot prohibit override redirect windows to become mapped.
+    fn mapped_override_redirect_window(&mut self, xwm: XwmId, window: X11Surface);
+    /// Window was unmapped.
+    fn unmapped_window(&mut self, xwm: XwmId, window: X11Surface);
+    /// Window was destroyed
+    fn destroyed_window(&mut self, xwm: XwmId, window: X11Surface);
```

Comment:
Is it garunteed that unmap would always be called before destroy? Or could destroy happen first because of a client disconnecting abruptly.

Reply:
I have never seen the XServer deliver these out of order.

Reply:
Perfect, that this is a non-issue.